File: src/ts/updateBrandBidAsk.ts
import "dotenv/config";
import { PrismaClient } from "@prisma/client";
export const prisma: PrismaClient = new PrismaClient();

/**
 * 銘柄・売買・値を読み取ってテーブルに登録する
 *
 * 実行方法は
 * npx ts-node src/updateBrandBidAsk.ts 銘柄名 bidかask 値(数値)
 */

// 引数チェック
if (process.argv.length !== 5) {
  console.error(
    `Error: Usage: npx ts-node src/updateBrandBidAsk.ts brand bid/ask (value). not (${process.argv})`
  );
  process.exit(1);
}

const brand = process.argv[2];
const bidAsk = process.argv[3];
const value = process.argv[4];

// 引数チェック
if (bidAsk !== "bid" && bidAsk !== "ask") {
  console.error("Error: bidAsk must be 'bid' or 'ask'. not " + bidAsk);
  process.exit(1);
}

// ファイル読み込み・解析
(async function updateBidAsk() {
  try {
    await prisma.brandBidAsk.upsert({
      where: {
        brand: brand.toUpperCase(),
      },
      update: {
        ...(bidAsk === "bid"
          ? {
              bid_price: +value,
              bid_updated_time: new Date(),
            }
          : {
              ask_price: +value,
              ask_updated_time: new Date(),
            }),
      },
      create: {
        brand: brand.toUpperCase(),
        ...(bidAsk === "bid"
          ? {
              bid_price: +value,
              ask_price: null,
              bid_updated_time: new Date(),
              ask_updated_time: null,
            }
          : {
              bid_price: null,
              ask_price: +value,
              bid_updated_time: null,
              ask_updated_time: new Date(),
            }),
      },
    });
  } catch (error) {
    console.error("データの登録に失敗しました:", error);
  }
})();

File: src/python/updateBrandBidAsk.py
import sys
import os
import os.path
from dotenv import load_dotenv
from supabase import create_client, Client
import datetime

def update_brand_bidask(brand,bidask,value):
    # .envファイルの内容を読み込見込む
    dotenv_path = os.path.join(os.path.dirname(__file__), '../../.env')
    load_dotenv(dotenv_path)

    # データベースとのコネクション確立
    url: str = os.environ.get("PYTHON_DB_URL")
    key: str = os.environ.get("PYTHON_DB_KEY")
    supabase: Client = create_client(url, key)

    if bidask == 'bid':
        supabase.table("brandBidAsk").upsert({"brand": brand, "bid_price": value, "bid_updated_time": datetime.datetime.now(datetime.timezone.utc).isoformat()}).execute()
    elif bidask == 'ask':
        supabase.table("brandBidAsk").upsert({"brand": brand, "ask_price": value, "ask_updated_time": datetime.datetime.now(datetime.timezone.utc).isoformat()}).execute()
    else:
        print("Error: arg2 is bid or ask. not ({0})".format(bidask))

if __name__ == "__main__": 
    if len(sys.argv) != 4:
        print("Error: Usage: {0} brand bid/ask (value). not ({0})".format(sys.argv))
        sys.exit(1)
    
    brand = sys.argv[1].upper()
    bidAsk = sys.argv[2]
    value = sys.argv[3]

    # 引数チェック
    if (bidAsk != "bid" and bidAsk != "ask"):
        print("Error: bidAsk must be 'bid' or 'ask'. not " + bidAsk)
        sys.exit(1)

    update_brand_bidask(brand,bidAsk,value) 
